### PR TITLE
Introduce `PINFuturePending` for handling delegate patterns

### PR DESCRIPTION
**Pull request by raycsh017**
This commit introduces `PINFuturePending`, a wrapper on top of `PINFuture`
that could be used to handle cases where we expect the future to be resolved
later in some other part of the code, e.g. 3rd party delegate method that
we have little control over when it will get called and resolve or reject the
future based on the result. The benefit of this is that we'll be able to chain
success/failure methods right away to the future without having to wait for it
to resolve or get rejected.

### CODE REVIEW

File: Example/Pods/Pods.xcodeproj/project.pbxproj
Comment (bolsinga) [inline Example/Pods/Pods.xcodeproj/project.pbxproj]: I have a hunch this needs to be made into a Public header file.

File: PINFuture/Classes/PINFuturePending.m
Comment (jparise) [inline PINFuture/Classes/PINFuturePending.m]: Is the `-copy` necessary if the properties are themselves marked as `copy`?

**Comment by jparise** [comment]
@raycsh017 do you think this should be mentioned in the documentation (`README.md`)?

**Comment by jparise** [comment]
@raycsh017 This could also use a unit test before we merge it.

**Comment by garrettmoon** [comment]
> @raycsh017 do you think this should be mentioned in the documentation (`README.md`)?

I second this, with a specific example of what this is good for.

File: PINFuture/Classes/PINFuturePending.m
Comment (jparise) [inline PINFuture/Classes/PINFuturePending.m]: It might be nicer to write this above the `if` (rather than in an `else` block) as:

```objc
NSAssert(self.resolve, @"Underlying future has already been resolved.");
```

... although the way you have it makes it easier to set a breakpoint on the error condition.

**Review by jparise — APPROVED** [review]

**Review by jparise — COMMENTED** [review]
It would still be nice to mention this in the `README.md`, too.

File: PINFuture/Classes/PINFuturePending.m
Comment (raycsh017) [reply PINFuture/Classes/PINFuturePending.m]: That sounds good. I just personally like having `else` cases to be clearer that we are handling all cases

**Comment by raycsh017** [comment]
@jparise @garrettmoon Updated `README.md` to include when and how to use `PINFuturePending`. Let me know if any part could be clearer :)

**Review by jparise — APPROVED** [review]
Looks good to me, but let's get one of the other folks to give this a 👍, too.

**Comment by raycsh017** [comment]
@garrettmoon @rahul-malik Could you guys take a look when you get the chance? Added documentation around the new class as well as unit tests!

File: PINFuture/Classes/PINFuturePending.m
Comment (rahul-malik) [inline PINFuture/Classes/PINFuturePending.m:53]: shouldn't the message be about rejection?

File: README.md
Comment (bolsinga) [inline README.md:285]: What if the PINFuture here was instead a PINFuture subclass, and it conformed to the GIDSIgnIn delegate itself? I'll contact you offline.

**Review by rahul-malik — APPROVED** [review]
@raycsh017 - Seems reasonable to me, the main concern I have is if the right solution here is another type or if we just need to create a new way without blocks for resolving / rejecting futures

I'll unblock this for now until I have a better solution in mind

File: PINFuture/Classes/PINFuturePending.m
Comment (raycsh017) [reply PINFuture/Classes/PINFuturePending.m:53]: Nope, since this could catch the case where you fulfill the pending future and try to call reject on the same future again.

**Pull request merged**
